### PR TITLE
Gem::Version should handle nil like it used to before

### PR DESCRIPTION
**Pull request by bronzdoc**
# Description:
closes https://github.com/rubygems/rubygems/issues/2359
reverts https://github.com/rubygems/rubygems/pull/2203

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

### CODE REVIEW

**Comment by duckinator** [comment]
:+1: if you add deprecation warnings for `nil` and empty strings saying they won't work starting with 4.x, as mentioned in the original issue.

**Comment by bronzdoc** [comment]
Oh shit i forgot that :sweat_smile: i will @duckinator thanks!

**Comment by tjouan** [comment]
>   * Gem::Version should handle  like it used to before

  Shouldn't we have "nil" between those two consecutive spaces?

**Comment by bronzdoc** [comment]
Since what was proposed was `I propose that we print a warning on nil in 3.x, and reject nil in 4.x. ` should we merge this and wait until 3.x to add the warning? @segiddins @duckinator

**Comment by segiddins** [comment]
Isn’t master now targeting 3.x?

File: lib/rubygems/version.rb [hunk 1]
```diff
@@ -170,6 +170,8 @@ def version
   # True if the +version+ string matches RubyGems' requirements.
 
   def self.correct? version
+    warn "nil versions are discouraged and will be deprecated in Rubygems 4" if version.nil?
+
     return false if version.nil?
```
Comment (hsbt) [inline lib/rubygems/version.rb]: Remove this line for reverting previous behavior.

Comment (bronzdoc) [reply lib/rubygems/version.rb]: done

**Comment by hsbt** [comment]
Ah, a test was failed caused by https://github.com/rubygems/rubygems/pull/2363/commits/41e1bc3640d9effd2ae04d9d72b8668036c9c054

**Comment by bronzdoc** [comment]
Fixed @hsbt

File: lib/rubygems/version.rb [hunk 2]
```diff
@@ -170,7 +170,8 @@ def version
   # True if the +version+ string matches RubyGems' requirements.
 
   def self.correct? version
-    return false if version.nil?
+    warn "nil versions are discouraged and will be deprecated in Rubygems 4" if version.nil?
```
Comment (segiddins) [inline lib/rubygems/version.rb]: should this use `Gem::Deprecate` ?

Comment (bronzdoc) [reply lib/rubygems/version.rb]: Should the year and month be? "2018", "12"

Comment (segiddins) [reply lib/rubygems/version.rb]: ah yeah, always forget that's how it works :/

In that case, only warn `unless Gem::Deprecate.skip?` ?

Comment (bronzdoc) [reply lib/rubygems/version.rb]: done

**Comment by bronzdoc** [comment]
@bundlerbot r+

**Pull request merged**